### ISSUE TITLE
Effectful, Futures and Android

### ISSUE THREAD

**Issue by stanch** [pelotom/effectful#8]
Hi,

I would like to apologize beforehand for somewhat lengthy question, seeking theoretical insights.
TL;DR: Is it guaranteed that everything inside `unwrap` is left as is and everything else is wrapped at least in `bind` and sometimes also in `pure`?

I am working with _Android_ platform, and there are two main concerns regarding threading:
- You can not perform network operations on _UI_ thread (results in an exception), otherwise long operations are discouraged as well, since they make the _UI_ slow.
- You can not perform _UI_ operations on threads other than _UI_. To overcome this, there is a `Handler` to which you can post `Runnable`s to be executed on _UI_ thread.

Running something on the _UI_ thread can be represented as a `Future`:

``` scala
@inline def runOnUiThread[A](f: ⇒ A): Future[A] = {
  val uiPromise = Promise[A]()
  if (uiThread == Thread.currentThread) {
    uiPromise.complete(Try(f))
  } else handler.post(new Runnable {
    def run() { uiPromise.complete(Try(f)) }
  })
  uiPromise.future
}
```

Next, we can also pimp `Future`s to handle the results on the _UI_ thread:

``` scala
implicit class RichFuture[A](val value: Future[A]) {
  def onSuccessUi(f: PartialFunction[A, Any])(implicit c: ExecutionContext) {
    value onSuccess { case v ⇒ runOnUiThread(f.lift(v)) }
  }
  def onFailureUi(f: PartialFunction[Throwable, Any])(implicit c: ExecutionContext) {
    value onFailure { case v ⇒ runOnUiThread(f.lift(v)) }
  }
}
...
future {
   getSomethingFromDb
} onSuccessUi { case v ⇒
   displayTheResults(v)
}
```

Now, with `akka-dataflow` we can define something like this (I am assuming you are familiar with the CPS plugin):

``` scala
@inline def switchToUiThread(): Unit @cps[Future[Any]] = shift { f: (Unit ⇒ Future[Any]) ⇒
  runOnUiThread(f())
}
...
flow {
  val v = future(getSomethingFromDb).apply() // same as await or unwrap
  switchToUiThread()
  displayTheResults(v)
} onFailureUi { case t ⇒ t.printStackTrace() }
```

I am quite happy with the `dataflow` approach, but it’s hard to use `@cps` in a loop, and `effectful` provides a more general solution, which I can use with `Option`s, etc (https://github.com/aztek/scala-workflow is nice too, but the future of untyped macros is unclear).

I have been fiddling with `effectful` for some time, and I think I have found a way to ensure that all _pure_ code is executed on _UI_ thread, whereas `unwrap`ped futures are mapped to a thread pool. Here it goes:

``` scala
implicit def futureUiMonad(implicit c: ExecutionContext) = new scalaz.Monad[Future] {
  override def point[A](a: ⇒ A): Future[A] = runOnUiThread(a)
  override def bind[A, B](fa: Future[A])(f: A ⇒ Future[B]): Future[B] = {
    val p = Promise[B]()
    fa onComplete {
      case f @ Failure(_) ⇒ p complete f.asInstanceOf[Failure[B]]
      case Success(a) ⇒ runOnUiThread(f(a)) onComplete {
        case f @ Failure(_) ⇒ p complete f.asInstanceOf[Failure[B]]
        case Success(b) ⇒ p completeWith b
      }
    }
    p.future
  }
}
```

Given that you have insight into the tree rewriting, I would be grateful if you could share your thoughts on whether I am correct about the results.

Kind regards,
Nick
